### ISSUE TITLE
HTTPS and isolates

### ISSUE THREAD

**Issue by spicemc** [stablekernel/aqueduct#229]
Hello,

I'm trying to use HTTPS and isolates. After creating a project with "aqueduct create" i added "bin/main.dart":

```dart
import 'dart:async';
import 'dart:io';

import 'package:aqueduct/aqueduct.dart';
import 'package:test_aqueduct/test_aqueduct.dart';

main() async {
  try {
    var certificateChain = Platform.script
        .resolve('certificates/server_chain.pem').toFilePath();
    var serverKey = Platform.script
        .resolve('certificates/server_key.pem').toFilePath();
    var securityContext = new SecurityContext();
    securityContext.useCertificateChain(certificateChain);
    securityContext.usePrivateKey(serverKey, password: 'dartdart');

    var app = new Application<TestAqueductSink>();
    var config = new ApplicationConfiguration()
      ..port = 8080
      ..configurationFilePath = "config.yaml"
      ..securityContext = securityContext;

    app.configuration = config;

    await app.start(numberOfInstances: 3); // fails
//    await app.start(runOnMainIsolate: true); // works
  } catch (e, st) {
    await writeError("$e\n $st");
  }
}

Future writeError(String error) async {
  print("$error");
}

```

I'm using IntelliJ to start the app:
`/usr/lib/dart/bin/dart --checked --enable-vm-service:43977 /path/to/bin/main.dart`

Error:
```
Invalid argument(s): Illegal argument in isolate message : (object extends NativeWrapper - Library:'dart:io' Class: _SecurityContext@14069316)
#0      Isolate._spawnFunction (dart:isolate-patch/isolate_patch.dart:474)
#1      Isolate.spawn.<spawn_async_body> (dart:isolate-patch/isolate_patch.dart:336)
#2      Future.Future.microtask.<anonymous closure> (dart:async/future.dart:184)
#3      _microtaskLoop (dart:async/schedule_microtask.dart:41)
#4      _startMicrotaskLoop (dart:async/schedule_microtask.dart:50)
#5      _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:96)
#6      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:149)
```

Now it is not possible to access the application. HTTPS and HTTP are not working, but the application is running somehow. If i use "runOnMainIsolate: true" it works, but with one isolate only. How is it possible to use HTTPS and many isolates?

**Comment by joeconwaystk** [comment]
Ah, I think I see what's up here, good catch. The security context can't be passed across isolates, and it naively is. The test coverage on using a security context is negligible, so this was not even tested (we typically host where the load balancer has the SSL cert, Dart app never sees it). 

I'm currently on vacation, don't have access to my dev machine. This may be a quick fix, but I'd like to give a little thought to the interface for providing SSL certs via the aqueduct serve CLI - if you have any suggestions before someone fixes this, let me know. In the meantime, sorry about that.

**Comment by spicemc** [comment]
Thanks for your quick reply. I have no suggestions right now and i`m looking forward to your solution. Btw it is a great piece of software, keep up the good work and enjoy your vacation.

**Comment by joeconwaystk** [comment]
I have a branch open that address this. Still have to write tests when I'm back in the office, so won't merge it into master for a bit. Can set pubspec.yaml ref to the following and run `pub get` to point at that branch:

```
aqueduct:
  git: https://github.com/stablekernel/aqueduct.git
  ref: jc/ssl
```

The change removes SecurityContext from the ApplicationConfiguration, instead there are two properties in ApplicationConfiguration for specifying the key/certificate path.

**Comment by spicemc** [comment]
I configured SSL using the new properties ```certificateFilePath``` and ```privateKeyFilePath```. Where can i set the password like ```securityContext.usePrivateKey(serverKey, password: 'dartdart');```?

**Comment by joeconwaystk** [comment]
OK, for configuration that requires more than just those two files, or in your case a password for the file, you may override the getter `securityContext` in your `RequestSink` to create the `SecurityContext` as you were before:

```
class TestAqueductSink extends RequestSink {
  @override 
  SecurityContext get securityContext {
   var certificateChain = Platform.script
        .resolve('certificates/server_chain.pem').toFilePath();
    var serverKey = Platform.script
        .resolve('certificates/server_key.pem').toFilePath();
    var ctx = new SecurityContext();
    ctx.useCertificateChain(certificateChain);
    ctx.usePrivateKey(serverKey, password: 'dartdart');
    return ctx;
  } 
  ..
}
```

Basically, the default implementation of this getter creates a context from the ApplicationConfiguration, which can be set via command-line args. Overriding this getter gets you finer grained control.

**Comment by spicemc** [comment]
Thank you, it is working as expected. Nice solution.

**Comment by joeconwaystk** [comment]
Now available on pub, v2.0.2.
